### PR TITLE
Add new cumulative_sum function to numpy and array_api namespaces

### PR DESCRIPTION
**Pull request by Micky774**
Towards #20200

Adds a new `cumulative_sum` function to the `jax.numpy` and `jax.experimental.array_api` namespaces in compliance with the array API 2023 standard.

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -5567,3 +5567,34 @@ def put(a: ArrayLike, ind: ArrayLike, v: ArrayLike,
   else:
     raise ValueError(f"mode should be one of 'wrap' or 'clip'; got {mode=}")
   return arr.at[unravel_index(ind_arr, arr.shape)].set(v_arr, mode=scatter_mode)
+
+
+@util.implements(getattr(np, 'cumulative_sum', None))
+def cumulative_sum(
+    x: Array, /, *, axis: int | None = None,
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5574]: We should use `ArrayLike` instead of `Array` for array arguments in `jax.numpy` APIs.

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -5567,3 +5567,34 @@ def put(a: ArrayLike, ind: ArrayLike, v: ArrayLike,
...
+    dtype: DTypeLike | None = None,
+    include_initial: bool = False) -> Array:
+  if isscalar(x) or x.ndim == 0:
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5577]: First validate input with `check_arraylike`, then convert to array with `asarray`

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -5567,3 +5567,34 @@ def put(a: ArrayLike, ind: ArrayLike, v: ArrayLike,
...
+    raise ValueError(
+      "The input must be non-scalar to take a cumulative sum, however a "
+      "scalar value or scalar array was given."
+    )
+  if axis is None and x.ndim > 1:
+    raise ValueError(
+      f"The input array has rank {x.ndim}, however axis was not set to an "
+      "explicit value. The axis argument is only optional for one-dimensional "
+      "arrays.")
+  axis = axis or 0
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5587]: Use `_canonicalize_axis` to validate the `axis` argument.

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -5567,3 +5567,34 @@ def put(a: ArrayLike, ind: ArrayLike, v: ArrayLike,
   else:
     raise ValueError(f"mode should be one of 'wrap' or 'clip'; got {mode=}")
   return arr.at[unravel_index(ind_arr, arr.shape)].set(v_arr, mode=scatter_mode)
+
+
+@util.implements(getattr(np, 'cumulative_sum', None))
+def cumulative_sum(
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5573]: Consider defining this function in `numpy/reductions.py`.

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -5567,3 +5567,34 @@ def put(a: ArrayLike, ind: ArrayLike, v: ArrayLike,
...
+  util.check_arraylike("cumulative_sum", x)
+  dtypes.check_user_dtype_supported(dtype)
+  kind = x.dtype.kind
+  if (dtype is None and kind in {'i', 'u'}
+      and x.dtype.itemsize*8 < int(config.default_dtype_bits.value)):
+    dtype = dtypes.dtype(dtypes._default_types[kind])
+
+  out = reductions.cumsum(x, axis=axis, dtype=dtype)
+  zeros_shape = list(x.shape)
+  zeros_shape[axis] = 1
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:5597]: Put this under the `if` statement?

File: tests/lax_numpy_test.py [hunk 6]
```diff
@@ -286,6 +286,59 @@ def np_fun(x):
...
+  @jtu.sample_product(
+    [dict(shape=shape, axis=axis)
+      for shape in all_shapes
+      for axis in list(range(-len(shape), len(shape))) + [None] if len(shape) == 1],
+    dtype=all_dtypes,
+    out_dtype=all_dtypes + [None],
+    include_initial=[False, True],
+  )
+  @jax.numpy_dtype_promotion('standard')  # This test explicitly exercises mixed type promotion
+  def testCumulativeSum(self, shape, axis, dtype, out_dtype, include_initial):
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:299]: Let's put these tests in `lax_numpy_reducers_test.py`

File: tests/lax_numpy_test.py [hunk 7]
```diff
@@ -286,6 +286,59 @@ def np_fun(x):
...
+    rng = jtu.rand_some_zero(self.rng())
+    x = rng(shape, dtype)
+    out = jnp.cumulative_sum(x, dtype=out_dtype, include_initial=include_initial)
+
+    target_dtype = out_dtype or x.dtype
+    kind = x.dtype.kind
+    if (out_dtype is None and kind in {'i', 'u'}
+        and x.dtype.itemsize*8 < int(config.default_dtype_bits.value)):
+      target_dtype = dtypes.dtype(dtypes._default_types[kind])
+    assert out.dtype == target_dtype
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:309]: We generally use `self._CheckAgainstNumpy` for these kinds of assertions. If `numpy` doesn't have the API yet, we can define a reference function with the correct semantics.

File: tests/lax_numpy_test.py [hunk 8]
```diff
@@ -286,6 +286,59 @@ def np_fun(x):
...
+    if include_initial:
+      target_shape[_axis] += 1
+    assert out.shape == tuple(target_shape)
+
+    target = jnp.cumsum(x, axis=_axis, dtype=out.dtype)
+    if include_initial:
+      zeros_shape = target_shape
+      zeros_shape[_axis] = 1
+      target = jnp.concat([jnp.zeros(target_shape, dtype=out.dtype), target])
+    self.assertArraysEqual(out, target)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:322]: Please also use `self._CompileAndCheck`, which asserts that the function's semantics don't change under JIT.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for working on this!

Also please add the new function to the docs by listing it here: https://github.com/google/jax/blob/main/docs/jax.numpy.rst

File: jax/_src/numpy/lax_numpy.py (hunk 4 above)
Comment (Micky774) [reply jax/_src/numpy/lax_numpy.py:5573]: I defined it here to avoid circular imports due to requiring `jax.numpy.concat`

Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:5573]: Maybe use `lax.concatenate` instead? `jnp.concat` is just a wrapper around this with some additional validation, which I don't think is needed in this case.

File: CHANGELOG.md [hunk 9]
```diff
@@ -11,6 +11,8 @@ Remember to align the itemized text with the first line of an item within a list
 * New Functionality
   * Added {func}`jax.numpy.unstack`, following the addition of this function in
     the array API 2023 standard, soon to be adopted by NumPy.
+  * Added {func}`jax.numpy.cumulative_sum`, following the addition of this
```
Comment (jakevdp) [inline CHANGELOG.md:14]: You could put both of these under one bullet point to keep the changelog more concise

File: tests/lax_numpy_reducers_test.py [hunk 10]
```diff
@@ -770,5 +770,72 @@ def test_f16_mean(self, dtype):
...
+    # input because we rely on JAX-specific casting behavior
+    args_maker = lambda: [jnp.array(rng(shape, dtype))]
+    np_op = getattr(np, "cumulative_sum", np_mock_fun)
+    kwargs = dict(axis=axis, dtype=out_dtype, include_initial=include_initial)
+    np_fun = lambda x: np_op(x, **kwargs)
+    jnp_fun = lambda x: jnp.cumulative_sum(x, **kwargs)
+    self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
+    self._CompileAndCheck(jnp_fun, args_maker)
+
+    kwargs = dict(axis=axis, include_initial=include_initial)
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py:818]: I'd avoid this kind of action at a distance (changing a global value in order to change the behavior of a previously-defined function). Maybe just include `out_dtype = None` in the parameterization, so you only have one pair of functions to check per test run?

File: tests/lax_numpy_test.py [hunk 11]
```diff
@@ -307,6 +307,7 @@ def np_fun(x):
                             atol={dtypes.bfloat16: 1e-1, np.float16: 1e-2})
     self._CompileAndCheck(jnp_fun, args_maker, atol={dtypes.bfloat16: 1e-1})
 
+
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:310]: revert

File: tests/lax_numpy_reducers_test.py [hunk 12]
```diff
@@ -770,5 +770,72 @@ def test_f16_mean(self, dtype):
...
+    return _dtype == dtypes.canonicalize_dtype(_dtype)
+
+  @jtu.sample_product(
+    [dict(shape=shape, axis=axis)
+      for shape in all_shapes
+      for axis in list(
+        range(-len(shape), len(shape))
+      ) + ([None] if len(shape) == 1 else [])],
+    dtype=filter(_is_canonical, all_dtypes),
+    out_dtype=filter(_is_canonical, all_dtypes),
```
Comment (jakevdp) [inline tests/lax_numpy_reducers_test.py:785]: I think the filter can be removed: I believe `all_dtypes` is defined in a way that makes this unnecessary.

File: tests/lax_numpy_reducers_test.py (hunk 10 above)
Comment (Micky774) [reply tests/lax_numpy_reducers_test.py]: > I'd avoid this kind of action at a distance (changing a global value in order to change the behavior of a previously-defined function)

I have updated the approach to avoid the indirect parameterization.

> Maybe just include `out_dtype = None` in the parameterization, so you only have one pair of functions to check per test run?

Because there are some important nuances with `dtype=None` based on various input dtypes, I think it is worth explicitly testing `dtype=None` _in each test instance_, otherwise w/ only 10 generated samples per test, even across CI there's a chance that we may have insufficient coverage. What do you think?

Comment (jakevdp) [reply tests/lax_numpy_reducers_test.py]: We run 90 samples per test internally, so I think it's probably OK.

File: jax/_src/numpy/reductions.py [hunk 13]
```diff
@@ -708,6 +708,40 @@ def _cumulative_reduction(a: ArrayLike, axis: Axis = None,
...
+    raise ValueError(
+      "The input must be non-scalar to take a cumulative sum, however a "
+      "scalar value or scalar array was given."
+    )
+  if axis is None and x.ndim > 1:
+    raise ValueError(
+      f"The input array has rank {x.ndim}, however axis was not set to an "
+      "explicit value. The axis argument is only optional for one-dimensional "
+      "arrays.")
+  axis = axis or 0
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py:728]: Can probably remove this
```suggestion
```

Comment (Micky774) [reply jax/_src/numpy/reductions.py:728]: `canonicalize_axis` doesn't handle `axis=None` though?

Comment (jakevdp) [reply jax/_src/numpy/reductions.py:728]: Oh, I saw the `ValueError` above and thought it was already handled.

If your intent is handling `axis=None`, then you should do so explicitly rather than setting `axis = 0` for any input with `bool(axis)` evaluating to False.

Comment (jakevdp) [reply jax/_src/numpy/reductions.py:728]: Maybe handle this as part of the `if` statement above?

**Review by jakevdp — APPROVED** [review]
Looks good!

File: jax/_src/numpy/reductions.py [hunk 14]
```diff
@@ -708,6 +708,42 @@ def _cumulative_reduction(a: ArrayLike, axis: Axis = None,
...
+        "explicit value. The axis argument is only optional for one-dimensional "
+        "arrays.")
+
+  axis = _canonicalize_axis(axis, x.ndim)
+  dtypes.check_user_dtype_supported(dtype)
+  kind = x.dtype.kind
+  if (dtype is None and kind in {'i', 'u'}
+      and x.dtype.itemsize*8 < int(config.default_dtype_bits.value)):
+    dtype = dtypes.canonicalize_dtype(dtypes._default_types[kind])
+  x = x.astype(dtype=dtype or x.dtype)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py:737]: Side-note: referencing the discussion in #20195, this is an example of a place where we wouldn't want `copy` in `astype` to default to `True`!
